File: vowpalwabbit/core/src/confidence_sequence.cc
// Copyright (c) by respective owners including Yahoo!, Microsoft, and
// individual contributors. All rights reserved. Released under a BSD (revised)
// license as described in the file LICENSE.

#if !defined(__APPLE__) && !defined(_WIN32)
#  define __STDCPP_MATH_SPEC_FUNCS__ 201003L
#  define __STDCPP_WANT_MATH_SPEC_FUNCS__ 1
#endif

#include "vw/core/estimators/confidence_sequence.h"

#include "vw/core/metric_sink.h"
#include "vw/core/model_utils.h"

#include <algorithm>
#include <cassert>
#include <cmath>

namespace VW
{
namespace details
{

incremental_f_sum incremental_f_sum::operator+=(double x)
{
  int i = 0;
  for (double y : this->partials)
  {
    if (std::abs(x) < std::abs(y)) { std::swap(x, y); }
    double hi = x + y;
    double lo = y - (hi - x);
    if (lo != 0.0)
    {
      this->partials[i] = lo;
      ++i;
    }
    x = hi;
  }
  this->partials.resize(i + 1);
  this->partials[i] = x;
  return *this;
}
incremental_f_sum incremental_f_sum::operator+(incremental_f_sum const& other) const
{
  incremental_f_sum result;
  result.partials = this->partials;
  for (double y : other.partials) { result += y; }
  return result;
}
details::incremental_f_sum::operator double() const
{
  double result = 0.0;
  for (double x : this->partials) { result += x; }
  return result;
}
}  // namespace details

namespace estimators
{
confidence_sequence::confidence_sequence(double alpha, double rmin_init, double rmax_init, bool adjust)
    : alpha(alpha), rmin_init(rmin_init), rmax_init(rmax_init), adjust(adjust)
{
  reset_stats();
}

void confidence_sequence::update(double w, double r, double p_drop, double n_drop)
{
  assert(w >= 0.0);
  assert(0.0 <= p_drop && p_drop < 1.0);
  assert(n_drop == -1.0 || n_drop >= 0.0);  // -1.0 equivalent to None in Python code

  ++update_count;
  last_w = w;
  last_r = r;

  if (!adjust) { r = std::min(rmax, std::max(rmin, r)); }
  else
  {
    rmin = std::min(rmin, r);
    rmax = std::max(rmax, r);
  }

  if (n_drop == -1.0) { n_drop = p_drop / (1.0 - p_drop); }

  double sumXlow = 0.0;   // NOLINT
  double Xhatlow = 0.0;   // NOLINT
  double sumXhigh = 0.0;  // NOLINT
  double Xhathigh = 0.0;  // NOLINT

  if (n_drop > 0.0)
  {
    sumXlow = (sumwr - sumw * rmin) / (rmax - rmin);
    double alow = sumXlow + 0.5;
    double blow = t + 1.0;
    sumxhatlowsq += std::pow(alow, 2) * (approxpolygammaone(blow) - approxpolygammaone(blow + n_drop));

    sumXhigh = (sumw * rmax - sumwr) / (rmax - rmin);
    double ahigh = sumXhigh + 0.5;
    double bhigh = t + 1.0;
    sumxhathighsq += std::pow(ahigh, 2) * (approxpolygammaone(bhigh) - approxpolygammaone(bhigh + n_drop));

    t += n_drop;
  }

  sumXlow = (sumwr - sumw * rmin) / (rmax - rmin);
  Xhatlow = (sumXlow + 0.5) / (t + 1.0);
  sumXhigh = (sumw * rmax - sumwr) / (rmax - rmin);
  Xhathigh = (sumXhigh + 0.5) / (t + 1.0);

  w /= (1.0 - p_drop);

  sumwsqrsq += std::pow(w * r, 2);
  sumwsqr += std::pow(w, 2) * r;
  sumwsq += std::pow(w, 2);
  sumwr += w * r;
  sumw += w;
  sumwrxhatlow += w * r * Xhatlow;
  sumwxhatlow += w * Xhatlow;
  sumxhatlowsq += std::pow(Xhatlow, 2);
  sumwrxhathigh += w * r * Xhathigh;
  sumwxhathigh += w * Xhathigh;
  sumxhathighsq += std::pow(Xhathigh, 2);

  ++t;
}

void confidence_sequence::persist(metric_sink& metrics, const std::string& suffix) const
{
  metrics.set_uint("upcnt" + suffix, update_count);
  metrics.set_float("lb" + suffix, lower_bound());
  metrics.set_float("ub" + suffix, upper_bound());
  metrics.set_float("last_w" + suffix, last_w);
  metrics.set_float("last_r" + suffix, last_r);
}

void confidence_sequence::reset_stats()
{
  rmin = rmin_init;
  rmax = rmax_init;
  eta = 1.1;
  s = 1.1;
  t = 0;

  sumwsqrsq.partials.clear();
  sumwsqr.partials.clear();
  sumwsq.partials.clear();
  sumwr.partials.clear();
  sumw.partials.clear();
  sumwrxhatlow.partials.clear();
  sumwxhatlow.partials.clear();
  sumxhatlowsq.partials.clear();
  sumwrxhathigh.partials.clear();
  sumwxhathigh.partials.clear();
  sumxhathighsq.partials.clear();

  update_count = 0;
  last_w = 0.0;
  last_r = 0.0;
}

float confidence_sequence::lower_bound() const
{
  if (t == 0 || rmin == rmax) { return static_cast<float>(rmin); }

  double sumvlow = (sumwsqrsq - 2.0 * rmin * sumwsqr + std::pow(rmin, 2) * sumwsq) / std::pow(rmax - rmin, 2) -
      2.0 * (sumwrxhatlow - rmin * sumwxhatlow) / (rmax - rmin) + sumxhatlowsq;
  double sumXlow = (sumwr - sumw * rmin) / (rmax - rmin);  // NOLINT
  double l = lblogwealth(sumXlow, sumvlow, eta, s, alpha / 2.0);

  return static_cast<float>(rmin + l * (rmax - rmin));
}

float confidence_sequence::upper_bound() const
{
  if (t == 0 || rmin == rmax) { return static_cast<float>(rmax); }

  double sumvhigh = (sumwsqrsq - 2.0 * rmax * sumwsqr + std::pow(rmax, 2) * sumwsq) / std::pow(rmax - rmin, 2) +
      2.0 * (sumwrxhathigh - rmax * sumwxhathigh) / (rmax - rmin) + sumxhathighsq;
  double sumXhigh = (sumw * rmax - sumwr) / (rmax - rmin);  // NOLINT
  double u = 1.0 - lblogwealth(sumXhigh, sumvhigh, eta, s, alpha / 2.0);

  return static_cast<float>(rmin + u * (rmax - rmin));
}

double confidence_sequence::approxpolygammaone(double b) const
{
  assert(b >= 1.0);
  if (b > 10.0)
  {
    double z = 1.0 / b;
    // Assuming[z > 0, FullSimplify@Series[PolyGamma[1, 1/z], { z, 0, 8 }]]
    return z * (1.0 + z * (0.5 + z * (1.0 / 6.0 + std::pow(z, 2) * (-1.0 / 30.0 + std::pow(z, 2) * (1.0 / 42.0)))));
  }
  else
  {
    // MiniMaxApproximation[PolyGamma[1, z], { z, { 1, 10 }, 4, 4 }]
    double numerator = 52.654073150313565 +
        b * (176.21984028201697 + b * (230.4407865623894 + b * (232.30138091080408 + b * (0.00006342628796017858))));
    double denominator =
        1.0 + b * (-7.709336904239235 + b * (80.59885744577618 + b * (114.25915616534267 + b * (232.30352278139097))));
    return numerator / denominator;
  }
}

double confidence_sequence::lblogwealth(double sumXt, double v, double eta, double s, double lb_alpha) const
{
#if !defined(__APPLE__) && !defined(_WIN32)
  double zeta_s = std::riemann_zeta(s);
#else
  double zeta_s = 10.584448464950803;  // std::riemann_zeta(s) -- Assuming s=1.1 is constant
#endif
  v = std::max(v, 1.0);
  double gamma1 = (std::pow(eta, 0.25) + std::pow(eta, 0.25)) / std::sqrt(2.0);
  double gamma2 = (std::sqrt(eta) + 1.0) / 2.0;
  assert(((std::log(eta * v) / std::log(eta)) + 1.0 > 0.0) && (1.0 + (std::log(eta * v) / std::log(eta)) != 0.0));
  double ll = s * std::log((std::log(eta * v) / std::log(eta)) + 1.0) + std::log(zeta_s / lb_alpha);
  return std::max(
      0.0, (sumXt - std::sqrt(std::pow(gamma1, 2) * ll * v + std::pow(gamma2, 2) * std::pow(ll, 2)) - gamma2 * ll) / t);
}
}  // namespace estimators

namespace model_utils
{
size_t read_model_field(io_buf& io, VW::details::incremental_f_sum& ifs)
{
  size_t bytes = 0;
  bytes += read_model_field(io, ifs.partials);
  return bytes;
}

size_t write_model_field(
    io_buf& io, const VW::details::incremental_f_sum& ifs, const std::string& upstream_name, bool text)
{
  size_t bytes = 0;
  bytes += write_model_field(io, ifs.partials, upstream_name + "_partials", text);
  return bytes;
}

size_t read_model_field(io_buf& io, VW::estimators::confidence_sequence& im)
{
  size_t bytes = 0;
  bytes += read_model_field(io, im.alpha);
  bytes += read_model_field(io, im.rmin_init);
  bytes += read_model_field(io, im.rmin);
  bytes += read_model_field(io, im.rmax_init);
  bytes += read_model_field(io, im.rmax);
  bytes += read_model_field(io, im.adjust);
  bytes += read_model_field(io, im.eta);
  bytes += read_model_field(io, im.s);
  bytes += read_model_field(io, im.t);
  bytes += read_model_field(io, im.sumwsqrsq);
  bytes += read_model_field(io, im.sumwsqr);
  bytes += read_model_field(io, im.sumwsq);
  bytes += read_model_field(io, im.sumwr);
  bytes += read_model_field(io, im.sumw);
  bytes += read_model_field(io, im.sumwrxhatlow);
  bytes += read_model_field(io, im.sumwxhatlow);
  bytes += read_model_field(io, im.sumxhatlowsq);
  bytes += read_model_field(io, im.sumwrxhathigh);
  bytes += read_model_field(io, im.sumwxhathigh);
  bytes += read_model_field(io, im.sumxhathighsq);
  bytes += read_model_field(io, im.update_count);
  bytes += read_model_field(io, im.last_w);
  bytes += read_model_field(io, im.last_r);
  return bytes;
}

size_t write_model_field(
    io_buf& io, const VW::estimators::confidence_sequence& im, const std::string& upstream_name, bool text)
{
  size_t bytes = 0;
  bytes += write_model_field(io, im.alpha, upstream_name + "_alpha", text);
  bytes += write_model_field(io, im.rmin_init, upstream_name + "_rmin_init", text);
  bytes += write_model_field(io, im.rmin, upstream_name + "_rmin", text);
  bytes += write_model_field(io, im.rmax_init, upstream_name + "_rmax_init", text);
  bytes += write_model_field(io, im.rmax, upstream_name + "_rmax", text);
  bytes += write_model_field(io, im.adjust, upstream_name + "_adjust", text);
  bytes += write_model_field(io, im.eta, upstream_name + "_eta", text);
  bytes += write_model_field(io, im.s, upstream_name + "_s", text);
  bytes += write_model_field(io, im.t, upstream_name + "_t", text);
  bytes += write_model_field(io, im.sumwsqrsq, upstream_name + "_sumwsqrsq", text);
  bytes += write_model_field(io, im.sumwsqr, upstream_name + "_sumwsqr", text);
  bytes += write_model_field(io, im.sumwsq, upstream_name + "_sumwsq", text);
  bytes += write_model_field(io, im.sumwr, upstream_name + "_sumwr", text);
  bytes += write_model_field(io, im.sumw, upstream_name + "_sumw", text);
  bytes += write_model_field(io, im.sumwrxhatlow, upstream_name + "_sumwrxhatlow", text);
  bytes += write_model_field(io, im.sumwxhatlow, upstream_name + "_sumwxhatlow", text);
  bytes += write_model_field(io, im.sumxhatlowsq, upstream_name + "_sumxhatlowsq", text);
  bytes += write_model_field(io, im.sumwrxhathigh, upstream_name + "_sumwrxhathigh", text);
  bytes += write_model_field(io, im.sumwxhathigh, upstream_name + "_sumwxhathigh", text);
  bytes += write_model_field(io, im.sumxhathighsq, upstream_name + "_sumxhathighsq", text);
  bytes += write_model_field(io, im.update_count, upstream_name + "_update_count", text);
  bytes += write_model_field(io, im.last_w, upstream_name + "_last_w", text);
  bytes += write_model_field(io, im.last_r, upstream_name + "_last_r", text);
  return bytes;
}
}  // namespace model_utils
}  // namespace VW

File: python/tests/confidence_sequence.py
from cmath import sqrt


class incremental_f_sum:
    """Incremental version of https://en.wikipedia.org/wiki/Kahan_summation_algorithm"""

    def __init__(self):
        self.partials = []

    def __iadd__(self, x):
        i = 0
        for y in self.partials:
            if abs(x) < abs(y):
                x, y = y, x
            hi = x + y
            lo = y - (hi - x)
            if lo:
                self.partials[i] = lo
                i += 1
            x = hi
        self.partials[i:] = [x]
        return self

    def __add__(self, other):
        result = incremental_f_sum()
        result.partials = deepcopy(self.partials)
        for y in other.partials:
            result += y
        return result

    def __float__(self):
        return sum(self.partials, 0.0)


class confidence_sequence(object):
    def __init__(self, rmin=0, rmax=1, adjust=True, eta=1.1, s=1.1):
        super().__init__()

        assert rmin <= rmax, (rmin, rmax)

        self.rho = 1
        self.eta = eta
        self.s = s
        self.rmin = rmin
        self.rmax = rmax
        self.adjust = adjust

        self.t = 0

        self.sumwsqrsq = incremental_f_sum()
        self.sumwsqr = incremental_f_sum()
        self.sumwsq = incremental_f_sum()
        self.sumwr = incremental_f_sum()
        self.sumw = incremental_f_sum()
        self.sumwrxhatlow = incremental_f_sum()
        self.sumwxhatlow = incremental_f_sum()
        self.sumxhatlowsq = incremental_f_sum()
        self.sumwrxhathigh = incremental_f_sum()
        self.sumwxhathigh = incremental_f_sum()
        self.sumxhathighsq = incremental_f_sum()

    def addobs(self, w, r, p_drop=0, n_drop=None):
        assert w >= 0
        assert 0 <= p_drop < 1
        assert n_drop is None or n_drop >= 0

        if not self.adjust:
            r = min(self.rmax, max(self.rmin, r))
        else:
            self.rmin = min(self.rmin, r)
            self.rmax = max(self.rmax, r)

        if n_drop is None:
            n_drop = p_drop / (1 - p_drop)

        if n_drop > 0:
            import scipy.special as sc

            # we have to simulate presenting n_drop events with w=0 in a row, which we can do in closed form
            # Sum[(a/(b + s))^2, { s, 0, n - 1 }]
            # a^2 PolyGamma[1,b]-a^2 PolyGamma[1,b+n]

            sumXlow = (float(self.sumwr) - float(self.sumw) * self.rmin) / (
                self.rmax - self.rmin
            )
            alow = sumXlow + 1 / 2
            blow = self.t + 1
            self.sumxhatlowsq += alow**2 * (
                sc.polygamma(1, blow).item() - sc.polygamma(1, blow + n_drop).item()
            )

            sumXhigh = (float(self.sumw) * self.rmax - float(self.sumwr)) / (
                self.rmax - self.rmin
            )
            ahigh = sumXhigh + 1 / 2
            bhigh = self.t + 1
            self.sumxhathighsq += ahigh**2 * (
                sc.polygamma(1, bhigh).item() - sc.polygamma(1, bhigh + n_drop).item()
            )

            self.t += n_drop

        sumXlow = (float(self.sumwr) - float(self.sumw) * self.rmin) / (
            self.rmax - self.rmin
        )
        Xhatlow = (sumXlow + 1 / 2) / (self.t + 1)
        sumXhigh = (float(self.sumw) * self.rmax - float(self.sumwr)) / (
            self.rmax - self.rmin
        )
        Xhathigh = (sumXhigh + 1 / 2) / (self.t + 1)

        w /= 1 - p_drop

        self.sumwsqrsq += (w * r) ** 2
        self.sumwsqr += w**2 * r
        self.sumwsq += w**2
        self.sumwr += w * r
        self.sumw += w
        self.sumwrxhatlow += w * r * Xhatlow
        self.sumwxhatlow += w * Xhatlow
        self.sumxhatlowsq += Xhatlow**2
        self.sumwrxhathigh += w * r * Xhathigh
        self.sumwxhathigh += w * Xhathigh
        self.sumxhathighsq += Xhathigh**2

        self.t += 1
        # print('r: ' + str(r) + ', w: ' + str(w))

    def getci(self, alpha):
        if self.t == 0 or self.rmin == self.rmax:
            return [self.rmin, self.rmax]

        sumvlow = (
            (
                float(self.sumwsqrsq)
                - 2 * self.rmin * float(self.sumwsqr)
                + self.rmin**2 * float(self.sumwsq)
            )
            / (self.rmax - self.rmin) ** 2
            - 2
            * (float(self.sumwrxhatlow) - self.rmin * float(self.sumwxhatlow))
            / (self.rmax - self.rmin)
            + float(self.sumxhatlowsq)
        )
        sumXlow = (float(self.sumwr) - float(self.sumw) * self.rmin) / (
            self.rmax - self.rmin
        )
        l = self.lblogwealth(
            t=self.t, sumXt=sumXlow, v=sumvlow, eta=self.eta, s=self.s, alpha=alpha / 2
        )

        sumvhigh = (
            (
                float(self.sumwsqrsq)
                - 2 * self.rmax * float(self.sumwsqr)
                + self.rmax**2 * float(self.sumwsq)
            )
            / (self.rmax - self.rmin) ** 2
            + 2
            * (float(self.sumwrxhathigh) - self.rmax * float(self.sumwxhathigh))
            / (self.rmax - self.rmin)
            + float(self.sumxhathighsq)
        )
        sumXhigh = (float(self.sumw) * self.rmax - float(self.sumwr)) / (
            self.rmax - self.rmin
        )
        u = 1 - self.lblogwealth(
            t=self.t,
            sumXt=sumXhigh,
            v=sumvhigh,
            eta=self.eta,
            s=self.s,
            alpha=alpha / 2,
        )
        return self.rmin + l * (self.rmax - self.rmin), self.rmin + u * (
            self.rmax - self.rmin
        )

    def lblogwealth(self, *, t, sumXt, v, eta, s, alpha):
        import scipy.special as sc
        from math import log, sqrt

        zeta_s = sc.zeta(s)
        v = max(v, 1)
        gamma1 = (eta ** (1 / 4) + eta ** (-1 / 4)) / sqrt(2)
        gamma2 = (sqrt(eta) + 1) / 2
        assert log(eta * v, eta) + 1 > 0, 1 + log(eta * v, eta)
        ll = s * log(log(eta * v, eta) + 1) + log(zeta_s / alpha)

        return max(
            0,
            (sumXt - sqrt(gamma1**2 * ll * v + gamma2**2 * ll**2) - gamma2 * ll)
            / t,
        )
